File: algorithms/cpp/77.组合.cpp
#include <iostream>
#include <vector>

using namespace std;
/*
 * @lc app=leetcode.cn id=77 lang=cpp
 *
 * [77] 组合
 */

// @lc code=start
class Solution {
 private:
  using return_type = vector<vector<int> >;
  void dfs(int n, int k, int cur, vector<int>& cur_ans, return_type& ans) {
    if (cur_ans.size() == k) {
      ans.push_back(cur_ans);
      return;
    }
    for (int i = cur; i < n + 1; i++) {
      cur_ans.push_back(i);
      dfs(n, k, i + 1, cur_ans, ans);
      cur_ans.pop_back();
    }
  }

 public:
  return_type combine(int n, int k) {
    return_type ans;
    vector<int> cur_ans;
    dfs(n, k, 1, cur_ans, ans);
    return ans;
  }
};
// @lc code=end

File: algorithms/golang/77.组合.go
/*
 * @lc app=leetcode.cn id=77 lang=golang
 *
 * [77] 组合
 */

// @lc code=start
/**
 * 递归
 */
func combine(n int, k int) [][]int {
	ans := [][]int{}
	if k == 0 {
		return ans
	}
	cur := []int{}
	var dfs func(i int)
	dfs = func(i int) {
		if len(cur) == k {
			comb := make([]int, k)
			copy(comb, cur)
			ans = append(ans, comb)
			return
		}
		if i > n {
			return
		}
		cur = append(cur, i)
		dfs(i + 1)
		cur = cur[:len(cur)-1]
		dfs(i + 1)
	}
	dfs(1)
	return ans
}

// @lc code=end